### PR TITLE
[MXNET-289] Fix bugs in image classification example 

### PR DESCRIPTION
**Pull request by rahul003**
## Description ##
https://issues.apache.org/jira/browse/MXNET-289

Fixes https://github.com/apache/incubator-mxnet/issues/9611

Problem 1:
The problem is that different parts of the dataset being worked on different machines can have different number of batches when the number of examples is not perfectly divisible. Some machines could get more one more batch than some other machines. On the last epoch, this causes issues as there's no progress because in dist_sync mode we wait for all workers to send gradients.

Using resizeIter to ensure each worker has same number of batches.

2. When using python3, polynomial learning rate crashes in previous code because epoch size is not integer
3.  Also when loading an epoch after all lr step changes have been done lr scheduler throws an error

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Updated image-classification examples to resize data iterator
- [x] Updated division in those examples to support both py2 and py3 by casting the result of division to int (now a/b can be float)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by rahul003** [comment]
@eric-haibin-lin @eftiquar please review

File: python/mxnet/module/base_module.py [hunk 1]
```diff
@@ -511,9 +514,12 @@ def fit(self, train_data, eval_data=None, eval_metric='acc',
                 self.forward_backward(data_batch)
                 self.update()
                 try:
-                    # pre fetch next batch
-                    next_data_batch = next(data_iter)
-                    self.prepare(next_data_batch, sparse_row_id_fn=sparse_row_id_fn)
+                    if epoch_size and (nbatch == epoch_size - 1):
```
Comment (marcoabreu) [inline python/mxnet/module/base_module.py]: Please correct me if I'm wrong, but this would be an infinite loop if the epoch_size was negative or 0, right? To be resilient against mis-usage (maybe coming from automated scripts), please assert ```epoch_size``` is positive and add a test accordingly

Comment (rahul003) [reply python/mxnet/module/base_module.py]: No, this is an argument that allows early stopping of loop in a way because different machines can have different number of samples. It can't run into an infinite loop.

Comment (rahul003) [reply python/mxnet/module/base_module.py]: This is also not really an easily testable scenario because the issue only comes up in distributed settings. The core logic just uses the argument provided. Rest of the changes in the PR is just for an example. I can add an assert sure

Comment (marcoabreu) [reply python/mxnet/module/base_module.py]: I see, thanks for elaborating :)

**Comment by anirudh2290** [comment]
@piiswrong can this be merged ?

File: python/mxnet/module/base_module.py [hunk 2]
```diff
@@ -399,7 +399,7 @@ def fit(self, train_data, eval_data=None, eval_metric='acc',
             eval_batch_end_callback=None, initializer=Uniform(0.01),
             arg_params=None, aux_params=None, allow_missing=False,
             force_rebind=False, force_init=False, begin_epoch=0, num_epoch=None,
-            validation_metric=None, monitor=None, sparse_row_id_fn=None):
+            validation_metric=None, monitor=None, sparse_row_id_fn=None, epoch_size=None):
```
Comment (piiswrong) [inline python/mxnet/module/base_module.py]: fit already has far too many arguments. Try using resizeiter

Comment (rahul003) [reply python/mxnet/module/base_module.py]: But we would still need to know the size of data iterator to determine number of batches to resize to

Comment (rahul003) [reply python/mxnet/module/base_module.py]: Oh you mean just relying on the user to use resizeIter?

**Comment by rahul003** [comment]
Removed the module changes and updated the example to show this can be resolved.

File: example/image-classification/common/fit.py [hunk 3]
```diff
@@ -155,9 +159,16 @@ def fit(args, network, data_loader, **kwargs):
     head = '%(asctime)-15s Node[' + str(kv.rank) + '] %(message)s'
     logging.basicConfig(level=logging.DEBUG, format=head)
     logging.info('start with arguments %s', args)
+    
+    epoch_size = get_epoch_size(args, kv)
 
     # data iterators
     (train, val) = data_loader(args, kv)
+    if 'dist' in args.kv_store and 'sync' in args.kv_store:
```
Comment (eric-haibin-lin) [inline example/image-classification/common/fit.py]: Is this also required for "async"?

Comment (rahul003) [reply example/image-classification/common/fit.py]: async should be good because no machine waits for other machines to pull gradients.

Comment (eric-haibin-lin) [reply example/image-classification/common/fit.py]: so there's no need to resize for async right?

**Comment by rahul003** [comment]
Yes, there's no need to do it in that case.

File: example/image-classification/common/fit.py (hunk 3 above)
Comment (eric-haibin-lin) [reply example/image-classification/common/fit.py]: Please check "async" instead

Comment (rahul003) [reply example/image-classification/common/fit.py]: Good catch Haibin, fixed it!

**Comment by rahul003** [comment]
@eric-haibin-lin @piiswrong Can we merge this?

**Comment by eric-haibin-lin** [comment]
CI didn’t pass

**Comment by rahul003** [comment]
@eric-haibin-lin The CI passed now, please take a look. Thanks!
